dcicutils/qa_utils.py

```diff
@@ -678,15 +682,487 @@ def Session(self, **kwargs):  # noQA - This name was chosen by AWS, so please do
+        if aws_access_key_id and aws_secret_access_key:
+            return Boto3Credentials(access_key=aws_access_key_id, secret_key=aws_secret_access_key)
+        return None
+
+    @property
+    def region_name(self) -> Optional[str]:
+        """
+        Returns the AWS region from the aws_region value, or in the credentials or config file
+        within the aws_credentials_dir, set in the constructor or set_credentials_for_testing();
+        or if not set there, then gets it via the standard AWS environment variable names,
+        i.e. AWS_REGION, AWS_DEFAULT_REGION, AWS_SHARED_CREDENTIALS_FILE, or AWS_CONFIG_FILE.
+
+        More specifically, returns AWS region from the first of these where defined; if defined returns None.
+        1. From the aws_region value set explicitly in the constructor or set_credentials_for_testing().
+        2. From the region property in the credentials file within the
+           aws_credentials_dir set explicitly in the constructor or set_credentials_for_testing().
+        3. From the value in the AWS_REGION environment variable (via os.environ).
+        4. From the value in the AWS_DEFAULT_REGION environment variable (via os.environ).
+        5. From the region property in the credentials file specified
+           by the AWS_SHARED_CREDENTIALS_FILE environment variable (via os.environ).
+        6. From the region property in the config file specified
+           by the AWS_CONFIG_FILE environment variable (via os.environ).
+        7. From the region property in the ~/.aws/credentials file.
+        8. From the region property in the ~/.aws/config file.
+
+        NOTE: Use unset_environ_credentials_for_testing() to clear related environment variables beforehand.
+
+        :return: AWS region determined as described above, or None.
+        """
+        aws_region = self.aws_region
+        if aws_region:
+            return aws_region
+        if self.aws_credentials_dir:
+            aws_credentials_file = os.path.join(self.aws_credentials_dir, "credentials")
+            _, _, aws_region = self._read_aws_credentials_from_file(aws_credentials_file)
+            if aws_region:
+                return aws_region
+            aws_config_file = os.path.join(aws_credentials_dir, "config")
+            _, _, aws_region = self._read_aws_credentials_from_file(aws_config_file)
+            if aws_region:
+                return aws_region
+        aws_region = os.environ.get("AWS_REGION", os.environ.get("AWS_DEFAULT_REGION"))
+        if aws_region:
+            return aws_region
+        _, _, aws_region = self._read_aws_credentials_from_file(os.environ.get("AWS_SHARED_CREDENTIALS_FILE", "~/.aws/credentials"))
+        if aws_region:
+            return aws_region
+        _, _, aws_region = self._read_aws_credentials_from_file(os.environ.get("AWS_CONFIG_FILE", "~/.aws/config"))
+        return aws_region
+
+
+# This MockBoto3Iam class is a minimal implementation, just enough to support the
```

Comment:
That's kind of true of all of these mocks. It's optional whether you actually say it, but it's fine if you do.